File: java-algo/src/main/java/com/leetcode/Problem718.java
package com.leetcode;

/**
 * 718. Maximum Length of Repeated Subarray
 * https://leetcode.com/problems/maximum-length-of-repeated-subarray/
 */
class Problem718 {
    public int findLength(int[] nums1, int[] nums2) {
        int m = nums1.length, n = nums2.length;
        int ans = 0;
        for (int i = 0; i < m; i++) {
            int len = Math.min(m - i, n);
            ans = Math.max(ans, getLength(nums1, i, nums2, 0, len));
        }
        for (int i = 0; i < n; i++) {
            int len = Math.min(m, n - i);
            ans = Math.max(ans, getLength(nums1, 0, nums2, i, len));
        }
        return ans;
    }

    private int getLength(int[] nums1, int x, int[] nums2, int y, int len) {
        int ret = 0, cur = 0;
        for (int i = 0; i < len; i++) {
            if (nums1[x + i] == nums2[y + i]) {
                if (++cur > ret)
                    ret = cur;
            } else {
                cur = 0;
            }
        }
        return ret;
    }
}

File: python-algo/leetcode/problem_718.py
"""718. Maximum Length of Repeated Subarray
https://leetcode.com/problems/maximum-length-of-repeated-subarray/

Given two integer arrays A and B, return the maximum length of an subarray
that appears in both arrays.

Example 1:

Input:
A: [1,2,3,2,1]
B: [3,2,1,4,7]
Output: 3
Explanation:
The repeated subarray with maximum length is [3, 2, 1].


Note:

1 <= len(A), len(B) <= 1000
0 <= A[i], B[i] < 100
"""
from typing import List


class Solution:
    def find_length(self, a: List[int], b: List[int]) -> int:
        """
        slide window
        """

        def cur_max_len(a_offset: int, b_offset: int, total_len: int) -> int:
            res = 0
            cur = 0
            for i in range(total_len):
                if a[a_offset + i] == b[b_offset + i]:
                    cur += 1
                    res = max(res, cur)
                else:
                    cur = 0
            return res

        a_len, b_len = len(a), len(b)
        ans = 0
        for i in range(a_len):
            length = min(a_len - i, b_len)
            ans = max(ans, cur_max_len(i, 0, length))
        for i in range(b_len):
            length = min(b_len - i, a_len)
            ans = max(ans, cur_max_len(0, i, length))
        return ans

    def find_length_1(self, a: List[int], b: List[int]) -> int:
        a_len, b_len = len(a), len(b)
        ans = 0
        # dp[i][j] means the subarray length which start a[i] and b[j]
        # dp[i][j] = dp[i+1][j+1] + 1 if a[i]==b[j] else 0
        dp = [[0] * (b_len + 1) for _ in range(a_len + 1)]
        for i in range(a_len - 1, -1, -1):
            for j in range(b_len - 1, -1, -1):
                if a[i] == b[j]:
                    dp[i][j] = dp[i + 1][j + 1] + 1
                    ans = max(ans, dp[i][j])
        return ans
